### PR TITLE
Fix make and makefile

### PR DESCRIPTION
**Pull request by matsduf**
CLI has the same problem as Engine with `share/Makefile` not being compatible with GNU Make on CentOS 7 (zonemaster/zonemaster-engine#813). The solution, as implemented by zonemaster/zonemaster-engine#814, made `share/Makefile` incompatible with BSD Make (on FreeBSD).

A partial solution to the problem for FreeBSD is proposed in zonemaster/zonemaster-engine#815.

This PR proposes a complete solution to the problem. GNU Make and BSD Make cannot sucessfully use the same Makefile, but GNU Make can be made available on FreeBSD as `gmake`. And for GNU Make the file `GNUmakefile` is the preferred file, if available. We use that fact.

* Update `Makefile` to be GNU Make compatible for CentOS 7 (and incompatible for BSD Make).
* Rename `Makefile` to `GNUmakefile`.
* Create wrapper `Makefile` for BSD Make to make it run the same target as `gmake` (and `GNUmakefile`).
* Update `Makefile.PL` to require `gmake` for FreeBSD and to use `gmake` for `share/GNUmakefile` under FreeBSD.
* Update installation instructions to include the installation of `gmake`.

The separation into two makefiles will make it possible to update those for future use.

### CODE REVIEW

**Comment by matsduf** [comment]
@sandoche2k, before you test build CLI on CentOS 7, review and test this PR.

File: Makefile.PL [hunk 1]
```diff
@@ -23,13 +23,30 @@ requires(
...
-	$(MYEXTLIB): share/Makefile
-	cd share && $(MAKE) all
-MAKE_FRAG
+    my $sharemakefile = 'share/GNUmakefile';
+    my $text;
+    if ($^O eq "freebsd") {
+        # Make FreeBSD use gmake for share
+        $text = q[GMAKE ?= "gmake"] . "\n"
+            . q[pure_all :: ] . "$sharemakefile\n"
+            . "\t" . q[cd share && $(GMAKE) all] . "\n";
```
Comment (mattias-p) [inline Makefile.PL:39]: Could you make the string quoting consistent? At the very least locally, but preferably also with the rest of the repo. E.g. `q[]` isn't used anywhere else in zonemaster-engine. Also, mixing concatenation and interpolation only makes it harder to read.

File: Makefile.PL [hunk 2]
```diff
@@ -23,13 +23,30 @@ requires(
...
+    } else {
+        # Here Linux and GNU Make is assumed
+        $text = q[pure_all :: ] . "$sharemakefile\n"
+            . "\t" . q[cd share && $(MAKE) all] . "\n";
+    };
```
Comment (mattias-p) [inline Makefile.PL:44]: Why not make both cases use the same variable for the make command?

**Review by mattias-p — CHANGES_REQUESTED** [review]
This is a quite non-trivial change. I don't think we should be merging this kind of thing this close to the release. It's better to adopt the same solution we used in Engine. We can do loftier things for the next release.

File: Makefile.PL (hunk 2 above)
Comment (matsduf) [reply Makefile.PL:44]: The $(MAKE) variable comes from the Makefile. The variable is interpolated by make.

File: Makefile.PL (hunk 1 above)
Comment (matsduf) [reply Makefile.PL]: Updated, but no change of logic.

**Comment by matsduf** [comment]
@mattias-p and @sandoche2k, please review again.

File: Makefile.PL (hunk 1 above)
Comment (matsduf) [reply Makefile.PL]: The changed has been tested for both FreeBSD and Ubuntu, and passed.

**Review by sandoche2k — APPROVED** [review]

File: Makefile.PL (hunk 2 above)
Comment (mattias-p) [reply Makefile.PL]: I think I would have gone for setting GMAKE to different values on different platforms. That way the "pure_all" clause could be the same for all platforms.

**Comment by mattias-p** [comment]
@matsduf I can't point to any technical errors in this PR, but do you really think we should make changes of this complexity this close to the release? We have much greater chances of finding problems with the Makefile-GNUmakefile approach if we merge it in the beginning of a release cycle. I think we should go for a more modest solution for this release.

**Comment by matsduf** [comment]
> We have much greater chances of finding problems with the Makefile-GNUmakefile approach if we merge it in the beginning of a release cycle.

It is well-documented that GNU Make prefers `GNUmakefile` before `Makefile`. The only problem would be the wrapper function, but that only affects FreeBSD which does not work well with the current situation anyway. I do not see that the risks are higher to implement than to keep it.

I am surprised that we have not seen the solution before instead of trying to make `Makefile` work for both GNU Make and BSD Make.

File: Makefile.PL (hunk 2 above)
Comment (matsduf) [reply Makefile.PL]: Such as "GMAKE = $(MAKE)" and "GMAKE = gmake", respectively?

Well, it has the same effect. Please the updated code.

**Comment by mattias-p** [comment]
It's true that GNU make will pick the GNUmakefile.

But it's no beauty queen. I guess that could be a reason why we haven't seen it before. For the next release I'll see if I can't come up with something more satisfying.

**Review by mattias-p — APPROVED** [review]

File: Makefile.PL (hunk 2 above)
Comment (mattias-p) [reply Makefile.PL]: Yeah, that's what I had in mind. I'm thinking that it's a little easier on the reader (i.e. us) if you don't have to compare two heavily quoted and generated makefile snippets by eye.

**Comment by matsduf** [comment]
Unless the makefile is very simple it does not seem to be possible to share it between BSD Make and GNU Make. I think this solution of having a wrapper for BSD Make is quite sound. Else we could leave `make` and use a shell script instead.

**Pull request merged**